lib/IR/AsmPrinter.cpp

```diff
@@ -1518,10 +1518,19 @@ void OperationPrinter::numberValueID(Value *value) {
   llvm::raw_svector_ostream specialName(specialNameBuffer);
 
   // Check to see if this value requested a special name.
-  auto *op = value->getDefiningOp();
-  if (state && op) {
-    if (auto *interface = state->getOpAsmInterface(op->getDialect()))
-      interface->getOpResultName(op, specialName);
+  if (state) {
+    if (auto *op = value->getDefiningOp()) {
+      // If the value is the result of an Op, ask the Op's dialect for a name
+      if (auto *interface = state->getOpAsmInterface(op->getDialect()))
+        interface->getOpResultName(op, specialName);
+    } else if (auto* arg = mlir::dyn_cast<mlir::BlockArgument>(value)) {
```

Comment:
This should be a cast. There are only two types of values.

Also, the hook should only be called on entry block arguments.

Also, drop all of the 'mlir::'.

Reply:
So I was doing a dyn_cast on the off chance there would some day be another subclass of Value, but I'm happy to change it to not do that.  Also, I will drop the pointless name-spacing, that was an oversight.  However, what is an 'entry' block argument?  A block argument on the first block in a region?  Is the idea that only those block arguments should be related to parent op's dialect?  Also, are non-entry block arguments even used anywhere?

Reply:
'entry' block argument is a block argument on the entry block of a region. We only really allow attaching semantics of any kind to these arguments.

Non-entry block arguments are used to represent phi values, which come into play in any place that has uses of non-structured control flow: llvm, standard, spirv, etc:
https://github.com/tensorflow/mlir/blob/master/g3doc/Rationale.md#block-arguments-vs-phi-nodes

Reply:
Thanks, we don't use much non-structured flow control, so I didn't realize this, but I love the use of block arguments instead of PHI nodes.  Should be fixed now.